### PR TITLE
222 grid memory issue

### LINKED ISSUE #222: The grid class requires a large amount of memory and time to set up
The grid class needs re-working to reduce memory requirements. At the moment it works and is fine for smaller parent and child grids but it doesn't scale well to large grids like UK500 and NAARC (which uses a masked global domain). This means these larger grids need unreasonable amounts of memory allocation on JASMIN (384Gb).

The solution is to start introducing xarray for reading in grid information. PyBDY only needs parent and child grid information along the boundaries. This is: 
  Parent: bdy +rim +9 surrounding points +2 for calculating grid variables e1, e2, glamu, levels with partial setp etc.
  Child: bdy +rim +2 for calculating grid variables. 
At the moment is gets grid information for the whole domain for both. With xarray we can work out where we will need grid information and then only extract it along required boundaries (this is similar to what pybdy does for src data with nemo_bdy_extr_assist.get_ind()).

### PR DESCRIPTION
**Pull request by b-barton**
Improvements have been made to the efficiency of pybdy for handling large grids. The improvements are detailed in #222. They include smaller memory usage and faster run time. This has been achieved by implementing the chunks on the grid objects. 
A couple of new tests have been added for the producing tidal boundary forcing files and to cover a few edge cases when the bdy points interact and get chunked with a complex coastline.

### CODE REVIEW

File: src/pybdy/tide/nemo_bdy_tide3.py [hunk 1]
```diff
@@ -51,293 +50,356 @@ def nemo_bdy_tide_rot(setup, DstCoord, Grid_T, Grid_U, Grid_V, comp):
     numharm = len(comp)
     logger = logging.getLogger(__name__)
     g_type = Grid_T.grid_type
-    DC = copy.deepcopy(DstCoord)
```
Comment (jdha) [inline src/pybdy/tide/nemo_bdy_tide3.py:54]: I'm finding it hard to work out the changes here as there are 250 lines of red then 250 of green - just to check you're happy removing the copy and that the before and after changes produce the same output? I vaguely remember having to put the deepcopy in years and years ago otherwise DstCoord was modified through DC ... but maybe that's not the case anymore.

**Review by jdha — COMMENTED** [review]
see comment in code - just need clarification

File: src/pybdy/tide/nemo_bdy_tide3.py (hunk 1 above)
Comment (b-barton) [reply src/pybdy/tide/nemo_bdy_tide3.py:54]: Yes I'm happy with removing the deepcopy. I added a regression test (test_tide()) of the tide code and tried it both before and after the changes. Perhaps I should have broken down the memory changes into an issue for the deepcopy and one for the chunk bit.

**Review by jdha — APPROVED** [review]
